### PR TITLE
add a simulation high level object and make serialsimulation not inherit...

### PR DESCRIPTION
**Pull request by JordiPolo**
... from roomba
- There is a new simulator class, wanted to call it simulation but we have a model with that name.  running on a thread.  So moving of the world and robots runs independently of the orders we get in the roomba interface(as it should be). This makes impossible to achieve some small movements. I am thinking on how to solve this.
- There is a new RobotSimulation class. Which is independent of the details of Roomba. 
- These 2 classes plus World and Console(poor's man logger) live in lib/simulation . I am thinking on moving them to a different gem. But at this moment is easier to modify everything in this project.
- RoombaSerialSimulation does not inherit from Roomba, and it can not touch the robot's positon directly. It's job is to get the basic commands of Roomba and calculate its velocity/turning, etc. Robot simulation and world will decide how the physics move Roomba.

Tests pass
The example of the readme works but end up in a different point. Must investigate this also.

### CODE REVIEW

**Comment by cwabbott** [comment]
Appears the simulated Roomba is stopping at a new X:Y and does not have a sensor reading:

`N: 90, x: 138, y: 110
Sensors: {:raw=>[0], :formatted=>"0"}
Sensors: {:raw=>[0], :formatted=>"0"}
Bytes Roomba received: [137, 0, 0, 0, 0]
Stopped moving
 => {:time=>1354797752, :bumps_and_drops=>{:raw=>[0], :formatted=>"0"}, :wall=>{:raw=>[0], :formatted=>"0"}}`

File: lib/roomba_serial_simulation.rb [hunk 1]
```diff
@@ -33,14 +44,67 @@ def write(*bytes)
...
+    ui['y'] = y
+    ui['radius'] = radius
+    ui['name'] = 'Roomba'
+    ui
+  end
+
+  def step(step_time)
+    if !@turning
+      distance = (@velocity * step_time).to_i
+      puts "Travelled #{distance}mm"
```
Comment (cwabbott) [inline lib/roomba_serial_simulation.rb]: Comment this out, not necessary when we have x:y coords.

File: lib/roomba_serial_simulation.rb [hunk 2]
```diff
@@ -33,14 +44,67 @@ def write(*bytes)
...
+      @previous_x = @x
+      @previous_y = @y
+      move_to(@facing, distance)
+      puts "N: #@facing, x: #@x, y: #@y"
```
Comment (cwabbott) [inline lib/roomba_serial_simulation.rb]: Something as basic as `puts "N: #@facing, x: #@x, y: #@y" unless @x == @previous_x && @y == @previous_y` will prevent a never ending thread of useless output to the console as the thread runs.

File: lib/roomba.rb [hunk 3]
```diff
@@ -131,15 +133,8 @@ def move(distance, degree=0, velocity=200)
...
-    # time = wheelbase * PI / 360degrees * degrees / velocity ABS
-    # wheelbase might be different for different roombas, consider refactoring
-    ((((ROOMBA_WHEELBASE * Math::PI) / 360) * degree.abs).to_f / velocity.to_f).abs
-  end
-
-  #spinning needs some work
-  def calculate_spin_degree(velocity, time)
-    ((time.to_f * velocity.to_f) / ((ROOMBA_WHEELBASE * Math::PI) / 360)) / 10**10
+  def current_time
+    Time.now
```
Comment (cwabbott) [inline lib/roomba.rb:139]: I don't get the purpose of this method. Should just call `Time.now` where we need it.

**Comment by JordiPolo** [comment]
@cwabbott I've done some extra and extensive changes to this PR

**Comment by cwabbott** [comment]
Interesting, I just glanced over it. Looks very different from the last version :dancer: 

Let's discuss at the meetup tonight :thumbsup:

**Comment by JordiPolo** [comment]
ch-ch-ch-changes  :rocket: 
I'll be there tonight

**Pull request closed without merging**